### ISSUE TITLE
Unstable Model Outputs, NaN Gradients with `fp16=True`

### ISSUE THREAD

**Issue by aaronhaefner** [aaronhaefner/applications-of-llm#22]
The model in the `app.py` `train_flan_t5` function is unstable and produces incoherent results when `fp16`, mixed precision training, is enabled. Gradients and evaluation metrics result in `NaN` values.

## Context

- **Environment:**
  - **Python Version:** 3.10
  - **Package Version:** Transformers 4.11.3, PyTorch 1.10

## Steps to Reproduce

1. Enable `fp16` in `TrainingArguments`:
   ```python
   training_args = TrainingArguments(
       ...
       fp16=True  # Enable mixed precision
   )
   ```

2. Run `python -m app` `train_flan_t5()` commented out

## Expected Behavior

The model should train successfully using `fp16`, with coherent loss values and non-NaN gradients.

## Actual Behavior

The model produces `NaN` values for gradients and evaluation loss. Example logs:
```
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 2e-05, 'epoch': 0.08}
{'eval_loss': nan, 'eval_runtime': 5.3046, 'eval_samples_per_second': 188.514, 'eval_steps_per_second': 47.129, 'epoch': 0.08}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 2e-05, 'epoch': 0.16}
{'eval_loss': nan, 'eval_runtime': 5.289, 'eval_samples_per_second': 189.071, 'eval_steps_per_second': 47.268, 'epoch': 0.16}
```

## Additional Information

Disabling `fp16` resolves the issue, with no NaN values observed during training.
